blockstore/autobatch.go

```diff
@@ -0,0 +1,98 @@
+package blockstore
+
+import (
+	"context"
+	"sync"
+
+	block "github.com/ipfs/go-block-format"
+	"github.com/ipfs/go-cid"
+)
+
+// autolog is a logger for the autobatching blockstore. It is subscoped from the
+// blockstore logger.
+var autolog = log.Named("auto")
+
+type AutobatchBlockstore struct {
+	bufferedBlks   []block.Block
+	addedCids      map[cid.Cid]struct{}
+	bufferedBlksLk sync.Mutex
+	flushLk        sync.Mutex
+	backingBs      Blockstore
+	bufferCapacity int
+	bufferSize     int
+}
+
+func (bs *AutobatchBlockstore) DeleteBlock(context.Context, cid.Cid) error {
+	panic("implement me")
+}
+
+func (bs *AutobatchBlockstore) Has(ctx context.Context, c cid.Cid) (bool, error) {
+	panic("implement me")
+}
+
+func (bs *AutobatchBlockstore) GetSize(context.Context, cid.Cid) (int, error) {
+	panic("implement me")
+}
+
+func (bs *AutobatchBlockstore) PutMany(context.Context, []block.Block) error {
+	panic("implement me")
+}
+
+func (bs *AutobatchBlockstore) AllKeysChan(ctx context.Context) (<-chan cid.Cid, error) {
+	panic("implement me")
+}
+
+func (bs *AutobatchBlockstore) HashOnRead(enabled bool) {
+	panic("implement me")
+}
+
+func (bs *AutobatchBlockstore) View(ctx context.Context, cid cid.Cid, callback func([]byte) error) error {
+	panic("implement me")
+}
+
+func (bs *AutobatchBlockstore) DeleteMany(ctx context.Context, cids []cid.Cid) error {
+	panic("implement me")
+}
+
+func NewAutobatch(ctx context.Context, backingBs Blockstore, bufferCapacity int) *AutobatchBlockstore {
+	bs := &AutobatchBlockstore{
+		backingBs:      backingBs,
+		bufferCapacity: bufferCapacity,
+		addedCids:      make(map[cid.Cid]struct{}),
+	}
+
+	return bs
+}
+
+// May NOT `Get` blocks that have been `Put` into this store
+// Only guaranteed to fetch those that were already in the backingBs at creation of this store and those at the most recent `Flush`
+func (bs *AutobatchBlockstore) Get(ctx context.Context, c cid.Cid) (block.Block, error) {
+	return bs.backingBs.Get(ctx, c)
+}
+
+func (bs *AutobatchBlockstore) Put(ctx context.Context, blk block.Block) error {
+	bs.bufferedBlksLk.Lock()
+	_, ok := bs.addedCids[blk.Cid()]
```

Comment:
Idk what the memory impact of this is yet -- I assume it helps perf, though it's fast enough without it.

If we wanted to ameliorate the `Get` may not return blocks that have been newly `Put`ted, we can use this to check _whether_ we should have the block, and then traverse `bs.bufferedBlks`. Atrocious perf, but might as well do it, since we're claiming it should never happen.